pallets/liquidity-proxy/src/lib.rs

```diff
@@ -471,101 +454,59 @@ impl<T: Config> Pallet<T> {
-                    },
-                ),
-        )
-        // Exchange aggregation
-        .fold(
-            (
-                SwapOutcome::new(balance!(0), OutcomeFee::new()),
-                Vec::new(),
-                Weight::zero(),
-            ),
-            |(mut outcome, mut sources, mut total_weight),
-             (swap_outcome, swap_sources, swap_weight)| {
-                outcome.amount = swap_outcome.amount;
-                outcome.fee = outcome.fee.merge(swap_outcome.fee);
-                merge_two_vectors_unique(&mut sources, swap_sources);
-                total_weight = total_weight.saturating_add(swap_weight);
-                Ok((outcome, sources, total_weight))
-            },
+                .map(|(from, to, swap_amount, cur_sender, cur_receiver)| -> Result<_, DispatchError> {
+                    let (swap_outcome, sources, weight) = Self::exchange_single(
+                        cur_sender,
+                        cur_receiver,
+                        &dex_info.base_asset_id,
+                        &from,
+                        &to,
+                        swap_amount.clone(),
+                        filter.clone(),
+                    )?;
+                    Ok((swap_outcome, sources, weight))
+                }),
         )
-    }
-
-    /// Calculate the input amount for a given `output_amount` for a sequence of direct swaps.
-    fn calculate_input_amount(
-        dex_info: &DEXInfo<T::AssetId>,
-        assets: &[T::AssetId],
-        output_amount: Balance,
-        filter: &LiquiditySourceFilter<T::DEXId, LiquiditySourceType>,
-    ) -> Result<(Balance, Weight), DispatchError> {
-        let mut amount = output_amount;
-        let mut total_weight = Weight::zero();
-
-        assets
-            .iter()
-            .rev()
-            .tuple_windows()
-            .map(|(to, from)| (from, to)) // Need to reverse pairs as well
-            .map(|(from, to)| -> Result<_, DispatchError> {
-                let (quote, _, _, weight) = Self::quote_single(
-                    &dex_info.base_asset_id,
-                    &from,
-                    &to,
-                    QuoteAmount::with_desired_output(amount),
-                    filter.clone(),
-                    true,
-                    true,
-                )?;
-                total_weight = total_weight.saturating_add(weight);
-                amount = quote.amount;
-                Ok(())
-            })
-            .for_each(drop);
-        Ok((amount, total_weight))
+            // Exchange aggregation
+            .fold(
+                (
+                    SwapOutcome::new(balance!(0), OutcomeFee::new()),
+                    Vec::new(),
+                    Weight::zero(),
+                ),
+                |(mut outcome, mut sources, mut total_weight),
+                 (swap_outcome, swap_sources, swap_weight)| {
+                    outcome.amount = swap_outcome.amount;
```

Comment:
I think you are wrong, `outcome.amount` must be the last outcome amount in the route.
For example, if `ExchangeRoute` has 2 swaps (here I skip desired in/out for simplicity):
```
(in: XSTUSD, out: XST, amount_in: 111, amount_out: 222), 
(in: XST, out: XOR, amount_in: 222, amount_out: 333)
```
Then `swap_outcome` after [`exchange_single`](https://github.com/sora-xor/sora2-network/pull/947/files#diff-f988a4abcacfa96d44157885a7a7c234fe425d18cd0536d094b85bb9907db38fR482) contains `amount_out`s:
```
(222, XST), (333, XOR)
```
We sum all the fees and weights as a result and take the last amount as the result. Thus, the result `(333, XOR)` indicates that the route changes 111 XST to  333 XOR.
If we sum outcome amounts, it would be 222 + 333 = 555 which is wrong.

Reply:
Ok